### PR TITLE
Gossipsub fixes and Initiator flagging fixes

### PR DESCRIPTION
**Pull request by sinkingsugar**
This PR solves very critical issues in gossipsub, specifically related to incoming connections and peers quality.

### CODE REVIEW

File: libp2p/stream/connection.nim [hunk 1]
```diff
@@ -33,6 +33,7 @@ type
     peerInfo*: PeerInfo
     observedAddr*: Multiaddress
     upgraded*: Future[void]
+    outbound*: bool
```
Comment (arnetheduck) [inline libp2p/stream/connection.nim:36]: why do we have both `dir` as a direction enum and outbound as a `bool`?

Comment (sinkingsugar) [reply libp2p/stream/connection.nim:36]: well as you can see from the diff I used to use `dir`, but that's not the actual wire direction.

Comment (arnetheduck) [reply libp2p/stream/connection.nim:36]: well, the point was more, why is it not using the direction enum?

also, I presume that what we want to capture here is whether or not the underlying connection is outbound or not, and not the mplex stream on top - this should be documented clearly along with a reason why / pointer to the spec, because it's hard to follow. 

finally, the name could be used to indicate some of this as well, ie `transportDirection` or something - conceptually, it goes a bit against the abstraction of channels and connections being layered - basically, once a transport is established, one doesn't "normally" care in what direction the underlying stream was established, or indeed how many layers of envelopes there are (in theory, on could run mplex inside mplex inside noise)

**Comment by sinkingsugar** [comment]
There are some peer counting issues, I might know where the issue is but I need to give it a test

File: libp2p/protocols/pubsub/pubsubpeer.nim [hunk 2]
```diff
@@ -189,9 +195,12 @@ proc connectImpl(p: PubSubPeer) {.async.} =
     # issue so we try to get a new on
     while true:
       await connectOnce(p)
-
+  except CancelledError:
+      raise
```
Comment (dryajov) [inline libp2p/protocols/pubsub/pubsubpeer.nim:199]: You probably want to drop the peer on cancellation as well.

**Comment by sinkingsugar** [comment]
Issues fixed, good to go now.

**Review by dryajov — APPROVED** [review]
LGTM
